### PR TITLE
Change capture_scalar_outputs to use SymInt/SymFloat rather than Tensor to model scalars

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93150

Previously, Dynamo faked support for item() when `capture_scalar_outputs` was True by representing it internally as a Tensor. With dynamic shapes, this is no longer necessary; we can represent it directly as a SymInt/SymFloat. Do so. Doing this requires you to use dynamic shapes; in principle we could support scalar outputs WITHOUT dynamic shapes but I won't do this unless someone hollers for it.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D42885775](https://our.internmc.facebook.com/intern/diff/D42885775)

### CODE REVIEW

**Comment by ezyang** [comment]
OK, so, uh, I can't actually land this, because although I can get the graph all the way to inductor, inductor doesn't know how to compile graphs with item calls. I guess I'll gate this behind another config as well.

**Review by albanD — COMMENTED** [review]
Sounds good to me but I'll let Voz take a look for the dynamo code

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42885775).

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -319,22 +319,16 @@ def call_method(
...
-            else:
-                unimplemented(f"Tensor.{name}")
+        elif name == "item" and not config.capture_scalar_outputs:
+            unimplemented(f"Tensor.{name}")
+        elif (
+            name == "item"
+            and config.capture_scalar_outputs
+            and not config.dynamic_shapes
+        ):
+            raise AssertionError(
```
Comment (eellison) [inline torch/_dynamo/variables/tensor.py:329]: Is it necessary to throw here ?

File: test/dynamo/test_export.py [hunk 2]
```diff
@@ -320,6 +320,7 @@ def func(x, z, k):
 
         self.assertTrue(torch._dynamo.utils.same(real_result, dynamo_result))
 
+    @patch.object(torch._dynamo.config, "dynamic_shapes", True)
```
Comment (voznesenskym) [inline test/dynamo/test_export.py:323]: this is already run under test/dynamo/test_dynamic_shapes.py, unless you want this to only run under dynamic shapes?

Comment (voznesenskym) [reply test/dynamo/test_export.py:323]: Saw rest of pr, sg

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:329]: Yes - misconfiguration should throw.

**Review by voznesenskym — APPROVED** [review]

File: test/dynamo/test_export.py (hunk 2 above)
Comment (ezyang) [reply test/dynamo/test_export.py:323]: There is a problem which is when we mark things as dynamic shapes True here, we actually are technically running the test twice but under same config. This is annoying to fix though, and is preexisting condition

**Pull request closed without merging**